gluster-exporter/metric_peer_counts.go

```diff
@@ -0,0 +1,181 @@
+	outBs, err := execCommand(cmdStr)
+	if err != nil {
+		return nil, err
+	}
+	dec := json.NewDecoder(bytes.NewReader(outBs))
+	var vgs []myVGDetails
+	var vgMap = make(map[string]myVGDetails)
+	for {
+		t, decErr := dec.Token()
+		if decErr == io.EOF {
+			break
+		}
+		if decErr != nil {
+			err = errors.New("Unable to parse JSON output: " + err.Error())
+			return nil, err
+		}
+		// if the token is 'vg', collect/decode the details into VGDetails array
+		if t == "vg" {
+			decErr = dec.Decode(&vgs)
+		}
+		if decErr != nil {
+			err = errors.New("JSON output changed, parse failed: " + err.Error())
+			return nil, err
+		}
+	}
+	for _, vg := range vgs {
+		// collect the unique vgs into the map
+		if _, ok := vgMap[vg.VGName]; !ok {
+			vgMap[vg.VGName] = vg
+		}
+	}
+	pMetrics := &PeerMetrics{VGCount: len(vgMap), LVMap: make(map[string]int)}
+	for k, vg := range vgMap {
+		log.Printf("%s: %+v\n", k, vg)
+		if count, convErr := strconv.Atoi(strings.TrimSpace(vg.LVCount)); convErr == nil {
+			pMetrics.LVMap[vg.VGName] = count
+		}
+		if count, convErr := strconv.Atoi(strings.TrimSpace(vg.PVCount)); convErr == nil {
+			pMetrics.PVCount += count
+		}
+	}
+	return pMetrics, nil
+}
+
+var (
+	gnrlMetricLabels = []MetricLabel{
+		{
+			Name: "name",
+			Help: "Name of the metric, whose count is collected (PV or LV or VG etc)",
+		},
+		{
+			Name: "peerID",
+			Help: "Peer ID of the host on which the metrics are collected",
+		},
+	}
+	// an additional information of 'vgName' is added
+	// this specifies which Volume Group the LV count belongs
+	lvMetricLabels = []MetricLabel{
+		{
+			Name: "name",
+			Help: "Logical Volume Metric details",
+		},
+		{
+			Name: "peerID",
+			Help: "Peer ID of the host on which the LV details are picked",
+		},
+		{
+			Name: "vgName",
+			Help: "Volume Group Name, shows no: of lvs under the specified Volume Group",
+		},
+	}
+
+	glusterPVCountMetric = newPrometheusGaugeVec(Metric{
+		Namespace: "gluster",
+		Name:      "pv_count",
+		Help:      "No: of physical volumes, got through pvs command",
+		LongHelp:  "",
+		Labels:    gnrlMetricLabels,
+	})
+	glusterLVCountMetric = newPrometheusGaugeVec(Metric{
+		Namespace: "gluster",
+		Name:      "lv_count",
+		Help:      "No: of logical volumes, got through lvs command",
+		LongHelp:  "",
+		Labels:    lvMetricLabels,
+	})
+	glusterVGCountMetric = newPrometheusGaugeVec(Metric{
+		Namespace: "gluster",
+		Name:      "vg_count",
+		Help:      "No: of volume groups, got through vgs command",
+		LongHelp:  "",
+		Labels:    gnrlMetricLabels,
+	})
+)
+
+func peerCounts() (err error) {
+	var peerID string
+	// 'gluster' is initialized inside 'main' function,
+	// so it is better to check whether it is available or not
+	if gluster != nil {
+		if peerID, err = gluster.LocalPeerID(); err != nil {
+			return
+		}
+	}
+	pMetrics, err := NewPeerMetrics()
+	if err != nil {
+		return err
+	}
+	log.Printf("Peer Metrics: %+v\n", pMetrics)
+	genrlLbls := prometheus.Labels{
```

Comment:
I dont think this way for creating labels for the metrics is correct. Rather I would define labels as below

```
vgLabels = []MetricLable{
    {Name: "peerid", Help: "Peer ID"}
}

lvLabels = []MetricLable{
    {Name: "peerid", Help: "Peer ID"},
    {Name: "vg", Help: "Volume Group name"}
}

pvLabels = []MetricLabels{
    {Name: "peerid", Help: "Peer ID"}
}
```

And I would add get functions for these labels separately with names `getPVLabels`, `getVGLabels` and `getPVLabels`

Reply:
Here you can see that, 'vgLabels' and 'pvLabels' are similar and there are other metrics which requires 'lvLabels' (where vg name is also included). So I tried to optimize the code by using 2 types of labels rather than per metric level separate labels.

'get' functions are required only if you have some calculations or modifications to be made, otherwise we could avoid a function call and use the label names directly.